### ISSUE TITLE
Application Crashing on Making Outbound Call 

### ISSUE THREAD

**Issue by deftsoftiphone** [twilio/voice-quickstart-ios#392]
### Description

Application is crashing when we start calling on any outbound call

### Steps to Reproduce

1. Registered with VOIP
2. Access Token Created
3. Call on outbound number
4. App got crashed

#### Code

```swift
2020-09-17 15:33:01.632029+0530 Second Phone Number[5369:237499] StartCallAction transaction request successful
2020-09-17 15:33:01.644056+0530 Second Phone Number[5369:237282] providerDidBegin
2020-09-17 15:33:01.644311+0530 Second Phone Number[5369:237282] provider:performStartCallAction:
2020-09-17 15:33:01.832713+0530 Second Phone Number[5369:237282] provider:didActivateAudioSession:
2020-09-17 15:33:02.602534+0530 Second Phone Number[5369:237545] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C9.1:2][0x10c740030] get output frames failed, state 8196
2020-09-17 15:33:02.602854+0530 Second Phone Number[5369:237545] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C9.1:2][0x10c740030] get output frames failed, state 8196
2020-09-17 15:33:02.603759+0530 Second Phone Number[5369:237545] TIC Read Status [9:0x0]: 1:57
2020-09-17 15:33:02.603864+0530 Second Phone Number[5369:237545] TIC Read Status [9:0x0]: 1:57
2020-09-17 15:33:02.610100+0530 Second Phone Number[5369:237545] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C10.1:2][0x10c547bb0] get output frames failed, state 8196
2020-09-17 15:33:02.610168+0530 Second Phone Number[5369:237545] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C10.1:2][0x10c547bb0] get output frames failed, state 8196
2020-09-17 15:33:02.611015+0530 Second Phone Number[5369:237545] TIC Read Status [10:0x0]: 1:57
2020-09-17 15:33:02.611044+0530 Second Phone Number[5369:237545] TIC Read Status [10:0x0]: 1:57
2020-09-17 15:33:03.401012+0530 Second Phone Number[5369:237282] callDidStartRinging:
2020-09-17 15:33:04.503996+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C12.1:2][0x10c54b210] get output frames failed, state 8196
2020-09-17 15:33:04.504094+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C12.1:2][0x10c54b210] get output frames failed, state 8196
2020-09-17 15:33:04.504316+0530 Second Phone Number[5369:237497] TIC Read Status [12:0x0]: 1:57
2020-09-17 15:33:04.504441+0530 Second Phone Number[5369:237497] TIC Read Status [12:0x0]: 1:57
2020-09-17 15:33:04.523239+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C13.1:2][0x10c79f1d0] get output frames failed, state 8196
2020-09-17 15:33:04.523357+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C13.1:2][0x10c79f1d0] get output frames failed, state 8196
2020-09-17 15:33:04.523920+0530 Second Phone Number[5369:237497] TIC Read Status [13:0x0]: 1:57
2020-09-17 15:33:04.524004+0530 Second Phone Number[5369:237497] TIC Read Status [13:0x0]: 1:57
2020-09-17 15:33:04.567849+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C14.1:2][0x10c564150] get output frames failed, state 8196
2020-09-17 15:33:04.567974+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C14.1:2][0x10c564150] get output frames failed, state 8196
2020-09-17 15:33:04.569970+0530 Second Phone Number[5369:237497] TIC Read Status [14:0x0]: 1:57
2020-09-17 15:33:04.570065+0530 Second Phone Number[5369:237497] TIC Read Status [14:0x0]: 1:57
2020-09-17 15:33:05.342497+0530 Second Phone Number[5369:237503] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C15.1:2][0x112013670] get output frames failed, state 8196
2020-09-17 15:33:05.342607+0530 Second Phone Number[5369:237503] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C15.1:2][0x112013670] get output frames failed, state 8196
2020-09-17 15:33:05.343111+0530 Second Phone Number[5369:237503] TIC Read Status [15:0x0]: 1:57
2020-09-17 15:33:05.343191+0530 Second Phone Number[5369:237503] TIC Read Status [15:0x0]: 1:57
2020-09-17 15:33:05.373982+0530 Second Phone Number[5369:237282] pushRegistry:didReceiveIncomingPushWithPayload:forType:completion:
2020-09-17 15:33:05.527470+0530 Second Phone Number[5369:237282] *** Terminating app due to uncaught exception 'TVOAudioDeviceMustBeSetBeforeMediaStackSetup', reason: 'WebRTC does not allow updating the audio device once the media stack is created (i.e once a track is created or memory allocated for previously created call or connect option is not deleted). You should provide the audio device to SDK prior to creating the tracks.'
*** First throw call stack:
(0x18e2d8518 0x18d4b39f8 0x18e1f5ec0 0x108df7450 0x108de4ce8 0x104a61b08 0x104a61b50 0x1ba2a68f8 0x1ba216d38 0x1ba217040 0x1ba217f7c 0x1ba219210 0x1ba1fc420 0x1bacd7e54 0x1927661f0 0x19276b198 0x1926ce0a8 0x1926fc108 0x1926fccf8 0x18e26989c 0x18e2645c4 0x18e264b40 0x18e264354 0x19046479c 0x1ba84fb68 0x104996bbc 0x18dd2a8e0)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

#### Expected Behavior

Call needs to work properly

#### Actual Behavior

App is crashing on connecting call

#### Reproduces How Often

100%

#### Twilio Call SID(s)

NT95737611adc715ede76fa2c28ffc7e7f
NT2ed458c0dec3cbed65ef9b598ef8f238

#### Logs

```
2020-09-17 15:33:01.632029+0530 Second Phone Number[5369:237499] StartCallAction transaction request successful
2020-09-17 15:33:01.644056+0530 Second Phone Number[5369:237282] providerDidBegin
2020-09-17 15:33:01.644311+0530 Second Phone Number[5369:237282] provider:performStartCallAction:
2020-09-17 15:33:01.832713+0530 Second Phone Number[5369:237282] provider:didActivateAudioSession:
2020-09-17 15:33:02.602534+0530 Second Phone Number[5369:237545] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C9.1:2][0x10c740030] get output frames failed, state 8196
2020-09-17 15:33:02.602854+0530 Second Phone Number[5369:237545] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C9.1:2][0x10c740030] get output frames failed, state 8196
2020-09-17 15:33:02.603759+0530 Second Phone Number[5369:237545] TIC Read Status [9:0x0]: 1:57
2020-09-17 15:33:02.603864+0530 Second Phone Number[5369:237545] TIC Read Status [9:0x0]: 1:57
2020-09-17 15:33:02.610100+0530 Second Phone Number[5369:237545] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C10.1:2][0x10c547bb0] get output frames failed, state 8196
2020-09-17 15:33:02.610168+0530 Second Phone Number[5369:237545] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C10.1:2][0x10c547bb0] get output frames failed, state 8196
2020-09-17 15:33:02.611015+0530 Second Phone Number[5369:237545] TIC Read Status [10:0x0]: 1:57
2020-09-17 15:33:02.611044+0530 Second Phone Number[5369:237545] TIC Read Status [10:0x0]: 1:57
2020-09-17 15:33:03.401012+0530 Second Phone Number[5369:237282] callDidStartRinging:
2020-09-17 15:33:04.503996+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C12.1:2][0x10c54b210] get output frames failed, state 8196
2020-09-17 15:33:04.504094+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C12.1:2][0x10c54b210] get output frames failed, state 8196
2020-09-17 15:33:04.504316+0530 Second Phone Number[5369:237497] TIC Read Status [12:0x0]: 1:57
2020-09-17 15:33:04.504441+0530 Second Phone Number[5369:237497] TIC Read Status [12:0x0]: 1:57
2020-09-17 15:33:04.523239+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C13.1:2][0x10c79f1d0] get output frames failed, state 8196
2020-09-17 15:33:04.523357+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C13.1:2][0x10c79f1d0] get output frames failed, state 8196
2020-09-17 15:33:04.523920+0530 Second Phone Number[5369:237497] TIC Read Status [13:0x0]: 1:57
2020-09-17 15:33:04.524004+0530 Second Phone Number[5369:237497] TIC Read Status [13:0x0]: 1:57
2020-09-17 15:33:04.567849+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C14.1:2][0x10c564150] get output frames failed, state 8196
2020-09-17 15:33:04.567974+0530 Second Phone Number[5369:237497] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C14.1:2][0x10c564150] get output frames failed, state 8196
2020-09-17 15:33:04.569970+0530 Second Phone Number[5369:237497] TIC Read Status [14:0x0]: 1:57
2020-09-17 15:33:04.570065+0530 Second Phone Number[5369:237497] TIC Read Status [14:0x0]: 1:57
2020-09-17 15:33:05.342497+0530 Second Phone Number[5369:237503] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C15.1:2][0x112013670] get output frames failed, state 8196
2020-09-17 15:33:05.342607+0530 Second Phone Number[5369:237503] [BoringSSL] nw_protocol_boringssl_get_output_frames(1301) [C15.1:2][0x112013670] get output frames failed, state 8196
2020-09-17 15:33:05.343111+0530 Second Phone Number[5369:237503] TIC Read Status [15:0x0]: 1:57
2020-09-17 15:33:05.343191+0530 Second Phone Number[5369:237503] TIC Read Status [15:0x0]: 1:57
2020-09-17 15:33:05.373982+0530 Second Phone Number[5369:237282] pushRegistry:didReceiveIncomingPushWithPayload:forType:completion:
2020-09-17 15:33:05.527470+0530 Second Phone Number[5369:237282] *** Terminating app due to uncaught exception 'TVOAudioDeviceMustBeSetBeforeMediaStackSetup', reason: 'WebRTC does not allow updating the audio device once the media stack is created (i.e once a track is created or memory allocated for previously created call or connect option is not deleted). You should provide the audio device to SDK prior to creating the tracks.'
*** First throw call stack:
(0x18e2d8518 0x18d4b39f8 0x18e1f5ec0 0x108df7450 0x108de4ce8 0x104a61b08 0x104a61b50 0x1ba2a68f8 0x1ba216d38 0x1ba217040 0x1ba217f7c 0x1ba219210 0x1ba1fc420 0x1bacd7e54 0x1927661f0 0x19276b198 0x1926ce0a8 0x1926fc108 0x1926fccf8 0x18e26989c 0x18e2645c4 0x18e264b40 0x18e264354 0x19046479c 0x1ba84fb68 0x104996bbc 0x18dd2a8e0)
libc++abi.dylib: terminating with uncaught exception of type NSException

```

### Versions

6.0.0

#### Voice iOS SDK

6.0.0

#### Xcode

11.6
#### iOS Version

12.2

#### iOS Device

iPhone 8 plus

**Comment by bobiechen-twilio** [comment]
Hi @deftsoftiphone 

With the CallKit related updates in the 6.0.0 release, the way audio device (TVOAudioDevice) is used needs to be slightly modified. Please see the [6.0.0 changelog](https://www.twilio.com/docs/voice/voip-sdk/ios/changelog#600) and more details in the [quickstart](https://github.com/twilio/voice-quickstart-ios). It would be great if you can try the quickstart with your environment and let us know if the SDK is still throwing the exception.

Thanks.

**Comment by deftsoftiphone** [comment]
Hi @bobiechen-twilio 
I have updated the Callkit with Version 6.0.0 but even after that, it is giving me the same error. I have updated all the audio device functions as per the new update.  Kindly look into the same

**Comment by bobiechen-twilio** [comment]
@deftsoftiphone if it's not too much trouble, could you try the quickstart with your credentials and let us know if the same crash happens in your environment?
Also does the exception happen even when making/accepting the first call after the app is launched?

**Comment by deftsoftiphone** [comment]
Hello @bobiechen-twilio 
it is giving us the same issue in quickstart as well and this is happening even when we try to call for the first time in the app

**Comment by bobiechen-twilio** [comment]
Thank for the update @deftsoftiphone.
I have tried the latest quickstart (we have recently updated the `master` branch to work with the 6.0 release) but was not able to get the exception.

Also from the SDK log it seems that you are making an outgoing call that will result in an incoming call to the local client. Is that correct? I tried the same case in quickstart and didn't get the exception but wanted to see if the exception is still thrown in your environment if you call another client with different identity.

-bobie

**Comment by deftsoftiphone** [comment]
@bobiechen-twilio 
I tried using the latest quickstart and updated the code it starts working now but now it creates a problem that ringing is going but Outbound call is not coming on other user end

**Comment by bobiechen-twilio** [comment]
Glad that is started working.

Please open a new issue with the debugging information gathered from the calls that you might have trouble connecting each others. From the description you might be experiencing some problem setting up the incoming capability on the callee client similar to some [previous issues](https://github.com/twilio/voice-quickstart-ios/issues?q=is%3Aissue+is%3Aclosed+incoming) opened by other SDK users.

Cheers.

**Comment by deftsoftiphone** [comment]
@bobiechen-twilio issue is again occuring, I just updated the twilio VOIP key and it start crashing again with same way

**Comment by bobiechen-twilio** [comment]
@deftsoftiphone 

The VoIP certificate should not have anything to do with the exception you are experiencing. Also, is the app making the outbound call to the identity registered on the same device?

**Comment by deftsoftiphone** [comment]
@bobiechen-twilio I am making outbound call on another device

**Comment by bobiechen-twilio** [comment]
@deftsoftiphone could you please provide the SDK log of the latest problem you are experiencing?
From the log of the initial description, there was a start-call action and an incoming-push-received callback before the exception was thrown. Please provide:
- Steps
- SDK log
- Call SID (might not be as helpful as the other information)

**Comment by deftsoftiphone** [comment]
Hello @bobiechen-twilio 

Steps are the same as mentioned in the details above

Below are the SDK logs:
```
2020-09-28 14:42:11.345952+0530 Second Phone Number[24001:5548481] providerDidBegin
2020-09-28 14:42:11.383247+0530 Second Phone Number[24001:5549785] StartCallAction transaction request successful
2020-09-28 14:42:11.386409+0530 Second Phone Number[24001:5548481] provider:performStartCallAction:
2020-09-28 14:42:11.674555+0530 Second Phone Number[24001:5548481] provider:didActivateAudioSession:
2020-09-28 14:42:12.598299+0530 Second Phone Number[24001:5548481] callDidStartRinging:
2020-09-28 14:42:14.768059+0530 Second Phone Number[24001:5548481] pushRegistry:didReceiveIncomingPushWithPayload:forType:completion:
2020-09-28 14:42:14.997472+0530 Second Phone Number[24001:5548481] *** Terminating app due to uncaught exception 'TVOAudioDeviceMustBeSetBeforeMediaStackSetup', reason: 'WebRTC does not allow updating the audio device once the media stack is created (i.e once a track is created or memory allocated for previously created call or connect option is not deleted). You should provide the audio device to SDK prior to creating the tracks.'
*** First throw call stack:
(0x1aeb2e5ac 0x1c2ba842c 0x1aea3962c 0x106ee7450 0x106ed4ce8 0x102bd0238 0x102bd0270 0x1b0d49678 0x1b0d4de14 0x1b0c91e38 0x1b0c92148 0x1b0c92fd0 0x1b0c94364 0x1b0c76690 0x1b1975dd4 0x1b1e7f280 0x1b1e7f73c 0x1b1e93b64 0x1b1ddb114 0x1b1e06418 0x1b1e07778 0x1aeaaa444 0x1aeaa4858 0x1aeaa4e08 0x1aeaa44bc 0x1c5529820 0x1b1448734 0x1b144de10 0x102b1201c 0x1ae76be60)
libc++abi.dylib: terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'TVOAudioDeviceMustBeSetBeforeMediaStackSetup', reason: 'WebRTC does not allow updating the audio device once the media stack is created (i.e once a track is created or memory allocated for previously created call or connect option is not deleted). You should provide the audio device to SDK prior to creating the tracks.'
terminating with uncaught exception of type NSException
```


Call SID: SID not generating now

**Comment by bobiechen-twilio** [comment]
Hi @deftsoftiphone 

As the exception message suggests, the application should not be assigning new audio device to the SDK while there is an active call session. Please update the app logic and make sure in your special use case the application is not reassigning audio device more than once.

**Comment by deftsoftiphone** [comment]
@bobiechen-twilio  this is happening for the first time making a call in which we assigning an audio device only once

**Comment by bobiechen-twilio** [comment]
Hi @deftsoftiphone 

Is it possible that the audio device is initialized and assigned to `TwilioVoice` when the outbound call was made but then similar `TwilioVoice.audioDevice` assignment happened again immediately when self-calling incoming call was received? Please keep in mind any audio device initialization or block execution need to happen before starting a call, either outbound or inbound. Manipulating audio device when there is an active call session will result in the exception.

**Comment by deftsoftiphone** [comment]
Hi @bobiechen-twilio 
I am initializing the Twilio Voice at the start of view controller and after that I am just enabling and disabling in the same way as mentioned in the demo, there is no change in the code for that. But it is still giving me the same issue

**Comment by bobiechen-twilio** [comment]
Hi @deftsoftiphone 

We have tried but cannot reproduce the exception using the quickstart. Please note that the audio device should be treated as a singleton in the app thus only one should be kept and managed, even across view controllers.

**Comment by deftsoftiphone** [comment]
Hi @bobiechen-twilio 
Please check the attached file of my code as I have done all the things as requested but it not worked
[CallingVC.swift.zip](https://github.com/twilio/voice-quickstart-ios/files/5332594/CallingVC.swift.zip)

**Comment by bobiechen-twilio** [comment]
Hi @deftsoftiphone 

What happens when the app receives an incoming call? does it use the same view controller or will a separate instance of the view controller created and presented?

Please note, as the exception message suggests, that the audio device should be configured prior to any inbound or outbound calls. Our suggestion here is to move the audio device (initialization) out of the view controller, in case it is created separately for outbound and inbound calls.

**Comment by deftsoftiphone** [comment]
Hi @bobiechen-twilio 
There is a single class for both incoming and outgoing call no separate class is there

**Comment by bobiechen-twilio** [comment]
@deftsoftiphone 

What I am trying to figure out is that whether view-controller is presented twice, one for the outgoing call and another time when the incoming call to the the mobile client itself is received. This case the audio device might be assigned more than once while there is still active call session.

**Comment by deftsoftiphone** [comment]
@bobiechen-twilio right now I am just opening a view controller only once when I am making a call. Also this issue is coming in outgoing call in which only single screen is there and not in case of incoming call. If we have  incoming call then it could be an issue but this issue occurs only in case of outgoing call

**Comment by bobiechen-twilio** [comment]
Hi @deftsoftiphone 

Thanks for the update. Please make sure any active call has ended before assigning the audio device instance.

**Comment by deftsoftiphone** [comment]
@bobiechen-twilio 
There is no active call in starting it is the first call I am initiating after running the app on phone

**Comment by bobiechen-twilio** [comment]
@deftsoftiphone 

Have you debugged how the `TwilioVoice.audioDevice` is accessed in your implementation? Could you please try the fresh quickstart without modification and attach SDK logs if the exception is still happening in your environment?

**Comment by deftsoftiphone** [comment]
@bobiechen-twilio I have checked with the latest code updated on git with no change and in this as well I am getting the same issue, please check the logs below as requested

`2020-10-15 12:21:36.675319+0530 Second Phone Number[19157:3999219] providerDidBegin
2020-10-15 12:21:38.051147+0530 Second Phone Number[19157:4000729] StartCallAction transaction request successful
2020-10-15 12:21:38.059461+0530 Second Phone Number[19157:3999219] provider:performStartCallAction:
2020-10-15 12:21:38.435657+0530 Second Phone Number[19157:3999219] provider:didActivateAudioSession:
2020-10-15 12:21:39.631700+0530 Second Phone Number[19157:3999219] callDidStartRinging:
2020-10-15 12:21:41.995249+0530 Second Phone Number[19157:3999219] pushRegistry:didReceiveIncomingPushWithPayload:forType:completion:
2020-10-15 12:21:42.052627+0530 Second Phone Number[19157:3999219] *** Terminating app due to uncaught exception 'TVOAudioDeviceMustBeSetBeforeMediaStackSetup', reason: 'WebRTC does not allow updating the audio device once the media stack is created (i.e once a track is created or memory allocated for previously created call or connect option is not deleted). You should provide the audio device to SDK prior to creating the tracks.'
*** First throw call stack:
(0x1824d25ac 0x19654c42c 0x1823dd62c 0x108967450 0x108954ce8 0x10464683c 0x10464652c 0x10464655c 0x1846ed678 0x1846f1e14 0x184635e38 0x184636148 0x184636fd0 0x184638364 0x18461a690 0x185319dd4 0x185823280 0x18582373c 0x185837b64 0x18577f114 0x1857aa418 0x1857ab778 0x18244e444 0x182448858 0x182448e08 0x1824484bc 0x198ecd820 0x184dec734 0x184df1e10 0x104586b3c 0x18210fe60)
libc++abi.dylib: terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'TVOAudioDeviceMustBeSetBeforeMediaStackSetup', reason: 'WebRTC does not allow updating the audio device once the media stack is created (i.e once a track is created or memory allocated for previously created call or connect option is not deleted). You should provide the audio device to SDK prior to creating the tracks.'
terminating with uncaught exception of type NSException
`

**Comment by bobiechen-twilio** [comment]
Hi @deftsoftiphone 

We still cannot reproduce this using the quickstart sample code. Please update your `viewDidLoad` method and only create and assign `TwilioVoice.audioDevice` once (or when necessary).

**Comment by deftsoftiphone** [comment]
@bobiechen-twilio I am using the same class from the quickstart sample code and just  updating my UI acordingly, rest code is same and no chage in that

**Comment by bobiechen-twilio** [comment]
@deftsoftiphone 

From the log you provided it looks evidently the audio device variable is assigned multiple time, which could be caused due to the two-calls use case in your app. Please use breakpoints to debug whether your application is setting audio device when you are handling the second call, the call that you make to yourself.

TwilioVoice.audioDevice should be a singleton across the app.

**Comment by deftsoftiphone** [comment]
Hello @bobiechen-twilio 
thanks it worked after that we have discussed the same with support of twilio on ticket. Issue was on swtting the TWIML which has been done now.
Thanks for your support
